File: datadome.go
package utils

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"
	"net/url"
	"os/exec"
	"visa/pkg/external/httpclient"
	"visa/pkg/external/httpclient/bogdanfinn"
)

type JSData struct {
	TTST     interface{} `json:"ttst"`
	IFOV     interface{} `json:"ifov"`
	HC       interface{} `json:"hc"`
	BrOh     interface{} `json:"br_oh"`
	BrOw     interface{} `json:"br_ow"`
	UA       interface{} `json:"ua"`
	WBD      interface{} `json:"wbd"`
	DP0      interface{} `json:"dp0"`
	TagPU    interface{} `json:"tagpu"`
	WDIF     interface{} `json:"wdif"`
	WDIFRM   interface{} `json:"wdifrm"`
	NPMTM    interface{} `json:"npmtm"`
	BrH      interface{} `json:"br_h"`
	BrW      interface{} `json:"br_w"`
	ISF      interface{} `json:"isf"`
	NDDC     interface{} `json:"nddc"`
	RsH      interface{} `json:"rs_h"`
	RsW      interface{} `json:"rs_w"`
	RsCd     interface{} `json:"rs_cd"`
	PHE      interface{} `json:"phe"`
	NM       interface{} `json:"nm"`
	JSF      interface{} `json:"jsf"`
	LG       interface{} `json:"lg"`
	PR       interface{} `json:"pr"`
	ArsH     interface{} `json:"ars_h"`
	ArsW     interface{} `json:"ars_w"`
	TZ       interface{} `json:"tz"`
	StrSs    interface{} `json:"str_ss"`
	StrLs    interface{} `json:"str_ls"`
	StrIdb   interface{} `json:"str_idb"`
	StrOdb   interface{} `json:"str_odb"`
	PLGOD    interface{} `json:"plgod"`
	PLG      interface{} `json:"plg"`
	PLGNE    interface{} `json:"plgne"`
	PLGRE    interface{} `json:"plgre"`
	PLGOF    interface{} `json:"plgof"`
	PLGGT    interface{} `json:"plggt"`
	PLTOD    interface{} `json:"pltod"`
	HCOVDR   interface{} `json:"hcovdr"`
	HCOVDR2  interface{} `json:"hcovdr2"`
	PLOVDR   interface{} `json:"plovdr"`
	PLOVDR2  interface{} `json:"plovdr2"`
	FTSOVDR  interface{} `json:"ftsovdr"`
	FTSOVDR2 interface{} `json:"ftsovdr2"`
	LB       interface{} `json:"lb"`
	EVA      interface{} `json:"eva"`
	LO       interface{} `json:"lo"`
	TsMtp    interface{} `json:"ts_mtp"`
	TsTec    interface{} `json:"ts_tec"`
	TsTsa    interface{} `json:"ts_tsa"`
	VND      interface{} `json:"vnd"`
	BID      interface{} `json:"bid"`
	MMT      interface{} `json:"mmt"`
	PLU      interface{} `json:"plu"`
	HDN      interface{} `json:"hdn"`
	AWE      interface{} `json:"awe"`
	GEB      interface{} `json:"geb"`
	DAT      interface{} `json:"dat"`
	MED      interface{} `json:"med"`
	ACO      interface{} `json:"aco"`
	ACOTS    interface{} `json:"acots"`
	ACMP     interface{} `json:"acmp"`
	ACMPTS   interface{} `json:"acmpts"`
	ACW      interface{} `json:"acw"`
	ACWTS    interface{} `json:"acwts"`
	ACMA     interface{} `json:"acma"`
	ACMATS   interface{} `json:"acmats"`
	ACA      interface{} `json:"acaa"`
	ACAATS   interface{} `json:"acaats"`
	AC3      interface{} `json:"ac3"`
	AC3TS    interface{} `json:"ac3ts"`
	ACF      interface{} `json:"acf"`
	ACFTS    interface{} `json:"acfts"`
	ACMP4    interface{} `json:"acmp4"`
	ACMP4TS  interface{} `json:"acmp4ts"`
	ACMP3    interface{} `json:"acmp3"`
	ACMP3TS  interface{} `json:"acmp3ts"`
	ACWM     interface{} `json:"acwm"`
	ACWMTS   interface{} `json:"acwmts"`
	OCPT     interface{} `json:"ocpt"`
	VCO      interface{} `json:"vco"`
	VCOTS    interface{} `json:"vcots"`
	VCH      interface{} `json:"vch"`
	VCHTS    interface{} `json:"vchts"`
	VCW      interface{} `json:"vcw"`
	VCWTS    interface{} `json:"vcwts"`
	VC3      interface{} `json:"vc3"`
	VC3TS    interface{} `json:"vc3ts"`
	VCMP     interface{} `json:"vcmp"`
	VCMPTS   interface{} `json:"vcmpts"`
	VCQ      interface{} `json:"vcq"`
	VCQTS    interface{} `json:"vcqts"`
	VC1      interface{} `json:"vc1"`
	VC1TS    interface{} `json:"vc1ts"`
	DVM      interface{} `json:"dvm"`
	SQT      interface{} `json:"sqt"`
	SO       interface{} `json:"so"`
	WDW      interface{} `json:"wdw"`
	COKYS    interface{} `json:"cokys"`
	ECPC     interface{} `json:"ecpc"`
	LGS      interface{} `json:"lgs"`
	LGSOD    interface{} `json:"lgsod"`
	PSN      interface{} `json:"psn"`
	EDP      interface{} `json:"edp"`
	ADDT     interface{} `json:"addt"`
	WSDC     interface{} `json:"wsdc"`
	CCSR     interface{} `json:"ccsr"`
	NUAD     interface{} `json:"nuad"`
	BCDA     interface{} `json:"bcda"`
	IDN      interface{} `json:"idn"`
	CAPI     interface{} `json:"capi"`
	SVDE     interface{} `json:"svde"`
	VPBQ     interface{} `json:"vpbq"`
	UCDV     interface{} `json:"ucdv"`
	SPWN     interface{} `json:"spwn"`
	EMT      interface{} `json:"emt"`
	BFR      interface{} `json:"bfr"`
	DBOV     interface{} `json:"dbov"`
	CFPFE    interface{} `json:"cfpfe"`
	STCFP    interface{} `json:"stcfp"`
	CKWA     interface{} `json:"ckwa"`
	PRM      interface{} `json:"prm"`
	TZP      interface{} `json:"tzp"`
	CVS      interface{} `json:"cvs"`
	USB      interface{} `json:"usb"`
	GLVD     interface{} `json:"glvd"`
	GLRD     interface{} `json:"glrd"`
	WWL      interface{} `json:"wwl"`
	JSET     interface{} `json:"jset"`
}

type Payload struct {
	DDV           interface{} `json:"ddv"`
	EventCounters interface{} `json:"eventCounters"`
	JSType        interface{} `json:"jsType"`
	DDK           interface{} `json:"ddk"`
	Request       interface{} `json:"request"`
	ResponsePage  interface{} `json:"responsePage"`
	CID           interface{} `json:"cid"`
	Referer       interface{} `json:"Referer"`
	JSData        JSData      `json:"jsData"`
}

type Headers struct {
	ContentType             string `json:"Content-type"`
	Host                    string `json:"Host"`
	Origin                  string `json:"Origin"`
	Referer                 string `json:"Referer"`
	AcceptEncoding          string `json:"Accept-Encoding"`
	AcceptLanguage          string `json:"Accept-Language"`
	SecChUa                 string `json:"Sec-Ch-Ua"`
	SecChUaMobile           string `json:"Sec-Ch-Ua-Mobile"`
	SecChUaPlatform         string `json:"Sec-Ch-Ua-Platform"`
	SecFetchDest            string `json:"Sec-Fetch-Dest"`
	SecFetchMode            string `json:"Sec-Fetch-Mode"`
	SecFetchSite            string `json:"Sec-Fetch-Site"`
	UserAgent               string `json:"User-Agent"`
	UpgradeInsecureRequests string `json:"Upgrade-Insecure-Requests"`
}

type JSONData struct {
	Headers Headers `json:"headers"`
	Payload Payload `json:"payload"`
}

type CookieData struct {
	Status int    `json:"status"`
	Cookie string `json:"cookie"`
}

func GetDataDome(ddv string, ddk string, urlStr string) (*JSONData, error) {
	if ddv == "" || ddk == "" {
		return nil, fmt.Errorf("ddv and ddk cannot be empty")
	}
	_, err := url.Parse(urlStr)
	if err != nil {
		return nil, fmt.Errorf("invalid url: %w", err)
	}

	binaryPath := "bin/datadome"
	cmd := exec.Command(binaryPath, "--ddv", ddv, "--ddk", ddk, "--url", urlStr)
	output, err := cmd.CombinedOutput()

	if err != nil {
		slog.Error("Error running DataDome binary:", "error", err)
		return nil, err
	}
	var datadome *JSONData
	jsonerr := json.Unmarshal(output, &datadome)
	if jsonerr != nil {
		slog.Error("Error parsing DataDome output:", "error", jsonerr)
		return nil, jsonerr
	}
	return datadome, nil
}

func GetDataDomeCookie(ddv string, ddk string, urlStr string) (string, error) {
	if ddv == "" || ddk == "" || urlStr == "" {
		return "", fmt.Errorf("ddv, ddk, and urlStr cannot be empty")
	}
	_, err := url.Parse(urlStr)
	if err != nil {
		return "", fmt.Errorf("invalid URL: %w", err)
	}
	datadome, err := GetDataDome(ddv, ddk, urlStr)
	if err != nil {
		slog.Error("Failed to get datadome cookie: ", "error", err)
		return "", err
	}
	client := bogdanfinn.NewStdClient()
	targetUrl := "https://api-js.datadome.co/js/"
	header := make(httpclient.EllisHeaders)
	header.Set("Content-type", datadome.Headers.ContentType)
	header.Set("Host", datadome.Headers.Host)
	header.Set("Origin", datadome.Headers.Origin)
	header.Set("Referer", datadome.Headers.Referer)
	header.Set("Accept-Encoding", datadome.Headers.AcceptEncoding)
	header.Set("Accept-Language", datadome.Headers.AcceptLanguage)
	header.Set("Sec-Ch-Ua", datadome.Headers.SecChUa)
	header.Set("Sec-Ch-Ua-Mobile", datadome.Headers.SecChUaMobile)
	header.Set("Sec-Ch-Ua-Platform", datadome.Headers.SecChUaPlatform)
	header.Set("Sec-Fetch-Dest", datadome.Headers.SecFetchDest)
	header.Set("Sec-Fetch-Mode", datadome.Headers.SecFetchMode)
	header.Set("Sec-Fetch-Site", datadome.Headers.SecFetchSite)
	header.Set("User-Agent", datadome.Headers.UserAgent)
	header.Set("Upgrade-Insecure-Requests", datadome.Headers.UpgradeInsecureRequests)

	// convert datadome.Payload.JSData to a JSON string
	jsDataJSon, err := json.Marshal(datadome.Payload.JSData)
	if err != nil {
		fmt.Println("Error marshaling payload:", err)
		return "", err
	}

	// build the request payload body
	body := url.Values{}
	body.Set("jsData", string(jsDataJSon))
	body.Add("ddv", datadome.Payload.DDV.(string))
	body.Add("jsType", datadome.Payload.JSType.(string))
	body.Add("ddk", datadome.Payload.DDK.(string))
	body.Add("request", datadome.Payload.Request.(string))
	body.Add("responsePage", datadome.Payload.ResponsePage.(string))
	body.Add("cid", datadome.Payload.CID.(string))
	body.Add("Referer", datadome.Payload.Referer.(string))

	response, err := client.Request(http.MethodPost, targetUrl, header, nil, bytes.NewBufferString(body.Encode()))
	if err != nil {
		slog.Error("Failed to request datadome url: ", "error", err)
		return "", err
	}
	defer response.Body.Close()

	if response.StatusCode != 200 {
		slog.Error("Failed to request datadome url: ", "error", response.StatusCode)
		return "", nil
	}

	var cookieResp CookieData
	err = json.NewDecoder(response.Body).Decode(&cookieResp)
	if err != nil {
		return "", err
	}
	// cookieResp.Cookie: datadome=GPv_HAoONj6YRdapOKu3xXgyicK3w3zYkewlAnvK43WXn61ctBzEMqQcIsBFOWlJk2UphTh~3E6HONYQcWOsGokKwT2DXDC8O_HBvPDTvor~j6~NIX4XsUGHj2Iz6mPs; Max-Age=31536000; Domain=.test.com; Path=/; Secure; SameSite=Lax
	return cookieResp.Cookie, nil
}

File: datadome.py
import random
import time
import base64

from browserforge.fingerprints import Screen, FingerprintGenerator


class DataDome:
    def __init__(self, ddv="4.25.0", ddk="your datadome key"):
        self.__ddv = ddv
        self.__ddk = ddk
        self.__screen = Screen(
            min_width=1280, max_width=5120, min_height=720, max_height=2880
        )
        self.__headers = FingerprintGenerator(
            browser=("chrome", "firefox", "safari", "edge"),
            os=("windows", "macos"),
            device="desktop",
            locale=("zh-CN", "en"),
            screen=self.__screen,
            http_version=2,
            strict=True,
            mock_webrtc=True,
        )
        self.__temps = self.__headers.generate()

    def build(self, url):
        stcfp = f"""ps://js.datadome.co/tags.js?id={self.__ddk}:2:90854)
    at https://js.datadome.co/tags.js?id={self.__ddk}:2:53225"""
        return {
            "headers": {
                "Content-type": "application/x-www-form-urlencoded",
                "Host": "api-js.datadome.co",
                "Origin": url,
                "Referer": url,
                "Accept-Encoding": self.__temps.headers.get("Accept-Encoding"),
                "Accept-Language": self.__temps.headers.get("Accept-Language"),
                "Sec-Ch-Ua": self.__temps.headers.get("sec-ch-ua"),
                "Sec-Ch-Ua-Mobile": self.__temps.headers.get("sec-ch-ua-mobile"),
                "Sec-Ch-Ua-Platform": self.__temps.headers.get("sec-ch-ua-platform"),
                "Sec-Fetch-Dest": "empty",
                "Sec-Fetch-Mode": "cors",
                "Sec-Fetch-Site": "cross-site",
                "User-Agent": self.__temps.headers.get("User-Agent"),
                "Upgrade-Insecure-Requests": self.__temps.headers.get(
                    "Upgrade-Insecure-Requests"
                ),
            },
            "payload": {
                "ddv": self.__ddv,
                "eventCounters": [],
                "jsType": "ch",
                "ddk": self.__ddk,
                "request": "%2F",
                "responsePage": "origin",
                "cid": "null",
                "Referer": url,
                "jsData": {
                    "ttst": f"{random.randint(10, 99)}.{random.randint(1000000000000, 9000000000000)}",
                    "ifov": "false",
                    "hc": self.__temps.navigator.hardwareConcurrency,
                    "br_oh": self.__temps.screen.outerHeight,
                    "br_ow": self.__temps.screen.outerWidth,
                    "ua": self.__temps.headers.get("User-Agent"),
                    "wbd": "false",
                    "dp0": "true",
                    "tagpu": f"{random.randint(10, 99)}.{random.randint(1000000000000, 9000000000000)}",
                    "wdif": "false",
                    "wdifrm": "false",
                    "npmtm": "false",
                    "br_h": self.__temps.screen.height,
                    "br_w": self.__temps.screen.width,
                    "isf": "false",
                    "nddc": 1,
                    "rs_h": self.__temps.screen.availHeight,
                    "rs_w": self.__temps.screen.availWidth,
                    "rs_cd": 24,
                    "phe": "false",
                    "nm": "false",
                    "jsf": "false",
                    "lg": self.__temps.navigator.language,
                    "pr": self.__temps.screen.devicePixelRatio,
                    "ars_h": self.__temps.screen.availHeight,
                    "ars_w": self.__temps.screen.availWidth,
                    "tz": -480,
                    "str_ss": "true",
                    "str_ls": "true",
                    "str_idb": "true",
                    "str_odb": "false",
                    "plgod": "false",
                    "plg": random.randint(5, 14),
                    "plgne": "true",
                    "plgre": "true",
                    "plgof": "false",
                    "plggt": "false",
                    "pltod": "false",
                    "hcovdr": "false",
                    "hcovdr2": "false",
                    "plovdr": "false",
                    "plovdr2": "false",
                    "ftsovdr": "false",
                    "ftsovdr2": "false",
                    "lb": "false",
                    "eva": 33,
                    "lo": "false",
                    "ts_mtp": 0,
                    "ts_tec": "false",
                    "ts_tsa": "false",
                    "vnd": self.__temps.navigator.vendor or "NA",
                    "bid": "NA",
                    "mmt": "application/pdf,text/pdf",
                    "plu": "PDF Viewer,Chrome PDF Viewer,Chromium PDF Viewer,Microsoft Edge PDF Viewer,WebKit built-in PDF",
                    "hdn": "false",
                    "awe": "false",
                    "geb": "false",
                    "dat": "false",
                    "med": "defined",
                    "aco": "probably",
                    "acots": "false",
                    "acmp": "probably",
                    "acmpts": "true",
                    "acw": "probably",
                    "acwts": "false",
                    "acma": "maybe",
                    "acmats": "false",
                    "acaa": "probably",
                    "acaats": "true",
                    "ac3": "",
                    "ac3ts": "false",
                    "acf": "probably",
                    "acfts": "false",
                    "acmp4": "maybe",
                    "acmp4ts": "false",
                    "acmp3": "probably",
                    "acmp3ts": "false",
                    "acwm": "maybe",
                    "acwmts": "false",
                    "ocpt": "false",
                    "vco": "",
                    "vcots": "false",
                    "vch": "probably",
                    "vchts": "true",
                    "vcw": "probably",
                    "vcwts": "true",
                    "vc3": "maybe",
                    "vc3ts": "false",
                    "vcmp": "",
                    "vcmpts": "false",
                    "vcq": "",
                    "vcqts": "false",
                    "vc1": "probably",
                    "vc1ts": "true",
                    "dvm": self.__temps.navigator.deviceMemory,
                    "sqt": "false",
                    "so": "landscape-primary",
                    "wdw": "true",
                    "cokys": "bG9hZFRpbWVzY3NpYXBwL=",
                    "ecpc": "false",
                    "lgs": "true",
                    "lgsod": "false",
                    "psn": "true",
                    "edp": "true",
                    "addt": "true",
                    "wsdc": "true",
                    "ccsr": "true",
                    "nuad": "true",
                    "bcda": "true",
                    "idn": "true",
                    "capi": "false",
                    "svde": "false",
                    "vpbq": "true",
                    "ucdv": "false",
                    "spwn": "false",
                    "emt": "false",
                    "bfr": "false",
                    "dbov": "false",
                    "cfpfe": "ZnVuY3Rpb24oKXt2YXIgdD1kb2N1bWVudFsnXHg3MVx4NzVceDY1XHg3Mlx4NzlceDUzXHg2NVx4NmNceDY1XHg2M1x4NzRceDZmXHg3MiddKCdceDYyXHg3Mlx4NmZceDc3XHg3M1x4NjVceDcyXHg2Nlx4NmNceDZmXHg3N1x4MmRceDYzXHg2Zlx4NmVceDc0XHg2",
                    "stcfp": base64.b64encode(stcfp.encode("utf-8")),
                    "ckwa": "true",
                    "prm": "true",
                    "tzp": "Asia/Shanghai",
                    "cvs": "true",
                    "usb": "defined",
                    "glvd": self.__temps.videoCard.vendor or "NA",
                    "glrd": self.__temps.videoCard.renderer or "NA",
                    "wwl": "false",
                    "jset": int(time.time()),
                },
            },
        }

    @staticmethod
    def cookies(cookie_str):
        parts = cookie_str.split("; ")
        cookie_dict = {}
        for part in parts:
            if "=" in part:
                key, value = part.split("=", 1)
                cookie_dict[key.lower()] = value

        cookie_key = "datadome"
        cookie_value = cookie_dict.get(cookie_key)
        cookie_domain = cookie_dict.get("domain")
        cookie_path = cookie_dict.get("path")

        return {
            "key": cookie_key,
            "value": cookie_value,
            "domain": cookie_domain,
            "path": cookie_path,
        }
